ampd/src/mvx/verifier.rs

```diff
@@ -0,0 +1,608 @@
+        if event_name.as_slice() != SIGNERS_ROTATED_EVENT.as_bytes() {
+            return Ok(false);
+        }
+
+        let signers_hash = topics.get(2).ok_or("")?;
+        let signers_hash = STANDARD.decode(signers_hash)?;
+
+        let weighted_signers = WeightedSigners::from(&self.verifier_set);
+
+        if signers_hash.len() != 32 || signers_hash.as_slice() != weighted_signers.hash().as_slice()
+        {
+            return Ok(false);
+        }
+
+        return Ok(true);
+    }
+}
+
+fn find_event<'a>(
+    transaction: &'a TransactionOnNetwork,
+    gateway_address: &Address,
+    log_index: usize,
+    identifier: &str,
+    needed_event_name: &[u8],
+) -> Option<&'a Events> {
+    if transaction.logs.is_none() {
+        return None;
+    }
+
+    // We only support log index being 0, so one supported event per transaction because of MultiversX limitations
+    if log_index != 0 {
+        return None;
+    }
+
+    // Because of current relayer limitation, if log_index is 0, we will search through the logs and get the first log
+    // which corresponds to the gateway address, hence only supporting one cross chain call per transaction
+
+    for event in transaction.logs.as_ref().unwrap().events.iter() {
+        if event.address.to_bytes() == gateway_address.to_bytes() && event.identifier == identifier
+        {
+            let topics = unwrap_or_continue!(event.topics.as_ref());
+
+            let event_name = unwrap_or_continue!(topics.get(0));
+            let event_name = STANDARD.decode(event_name).unwrap_or(Vec::new());
+            if event_name.as_slice() == needed_event_name {
+                return Some(event);
+            }
+        }
+    }
+
+    return None;
+}
+
+pub fn verify_message(
+    gateway_address: &Address,
+    transaction: &TransactionOnNetwork,
+    message: &Message,
+) -> Vote {
+    match find_event(
+        transaction,
+        gateway_address,
+        message.event_index as usize,
+        CONTRACT_CALL_IDENTIFIER,
+        CONTRACT_CALL_EVENT.as_bytes(),
+    ) {
+        Some(event)
+        if transaction.hash.as_ref().unwrap() == &message.tx_id.encode_hex::<String>()
+            && message.eq_event(event).unwrap_or(false) =>
+            {
+                Vote::SucceededOnChain
+            }
+        _ => Vote::NotFound,
+    }
+}
+
+pub fn verify_verifier_set(
+    gateway_address: &Address,
+    transaction: &TransactionOnNetwork,
+    verifier_set: VerifierSetConfirmation,
+) -> Vote {
+    match find_event(
+        transaction,
+        gateway_address,
+        verifier_set.event_index as usize,
+        ROTATE_SIGNERS_IDENTIFIER,
+        SIGNERS_ROTATED_EVENT.as_bytes(),
+    ) {
+        Some(event)
+        if transaction.hash.as_ref().unwrap() == &verifier_set.tx_id.encode_hex::<String>()
+            && verifier_set.eq_event(event).unwrap_or(false) =>
+            {
+                Vote::SucceededOnChain
+            }
+        _ => Vote::NotFound,
+    }

```

Comment:
no `unwrap()` in production code

```suggestion
let hash = transaction
        .hash
        .as_ref()
        .map(String::as_str)
        .unwrap_or_default();

    if hash.is_empty() {
        return Vote::NotFound;
    }

    match find_event(
        transaction,
        gateway_address,
        verifier_set.event_index as usize,
        ROTATE_SIGNERS_IDENTIFIER,
        SIGNERS_ROTATED_EVENT.as_bytes(),
    ) {
        Some(event)
            if hash == verifier_set.tx_id.encode_hex::<String>().as_str()
                && verifier_set.eq_event(event).unwrap_or(false) =>
        {
            Vote::SucceededOnChain
        }
        _ => Vote::NotFound,
    }
```

Reply:
Done.